### PR TITLE
Hughsie/udev tests prep1

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
I have no concerns with what's been pushed so far!

File: libfwupdplugin/fu-udev-device.h [hunk 1]
```diff
@@ -112,10 +62,6 @@ void
 fu_udev_device_set_bind_id(FuUdevDevice *self, const gchar *bind_id) G_GNUC_NON_NULL(1);
 const gchar *
 fu_udev_device_get_driver(FuUdevDevice *self) G_GNUC_NON_NULL(1);
-gboolean
-fu_udev_device_is_pci_base_cls(FuUdevDevice *self, FuPciBaseCls cls) G_GNUC_NON_NULL(1);
-guint32
-fu_udev_device_get_cls(FuUdevDevice *self) G_GNUC_NON_NULL(1);
```
Comment (superm1) [inline libfwupdplugin/fu-udev-device.h:118]: these should probably be _mentioned_ in migration guide

File: libfwupdplugin/fu-device.h [hunk 2]
```diff
@@ -640,6 +640,15 @@ fu_device_new(FuContext *ctx);
  */
 #define FU_DEVICE_PRIVATE_FLAG_SKIPS_RESTART "skips-restart"
 
+/**
+ * FU_DEVICE_PRIVATE_FLAG_IS_FAKE:
+ *
+ * The device is fake, and constructed from a fake sysfs test directory.
```
Comment (superm1) [inline libfwupdplugin/fu-device.h:646]: can't use emulated instead?

File: plugins/scsi/fu-scsi-device.c [hunk 3]
```diff
@@ -173,6 +173,14 @@ fu_scsi_device_probe(FuDevice *device, GError **error)
 		}
 	}
 
+	/* fake something as we cannot use ioctls */
+	if (fu_device_has_private_flag(device, FU_DEVICE_PRIVATE_FLAG_IS_FAKE)) {
+		fu_device_add_instance_str(device, "VEN", "ACME");
```
Comment (superm1) [inline plugins/scsi/fu-scsi-device.c:178]: what would you think about using `FWUPD`, `LINUX` or `TEST` for the fake vendor ID instead?  Just don't want to some day run into a problem that we have a "real" ACME some day.

File: src/fu-udev-backend.c [hunk 4]
```diff
@@ -335,6 +392,63 @@ fu_udev_backend_uevent_cb(GUdevClient *gudev_client,
...
+	const gchar *basename;
+	g_autoptr(GDir) dir = NULL;
+	g_autoptr(GError) error_dir = NULL;
+
+	dir = g_dir_open(fn, 0, &error_dir);
+	if (dir == NULL) {
+		if (g_error_matches(error_dir, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+			return;
+		g_debug("ignoring: %s", error_dir->message);
+		return;
```
Comment (superm1) [inline src/fu-udev-backend.c:407]: you can drop an extra return statement if you swap it like this:

```suggestion
		if (!g_error_matches(error_dir, G_FILE_ERROR, G_FILE_ERROR_NOENT))
			g_debug("ignoring: %s", error_dir->message);
		return;
```

**Review by superm1 — DISMISSED** [review]
I think it would be good to add some documentation on how people are "expected" to use this or a helper script that filters out stuff we don't care about in upstreaming.

File: libfwupdplugin/fu-udev-device.h (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-udev-device.h:118]: Good idea, fixed.

File: libfwupdplugin/fu-device.h (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-device.h:646]: Exactly that -- i.e. we create them enough for them to be added to the daemon as valid devices, but we can't interact with them in any way.

File: plugins/scsi/fu-scsi-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/scsi/fu-scsi-device.c:178]: 100% valid -- fixed, thanks.

File: src/fu-udev-backend.c (hunk 4 above)
Comment (hughsie) [reply src/fu-udev-backend.c:407]: Agreed, i've used your suggestion, thanks.

**Pull request merged**